File: solutions/1209-design-bounded-blocking-queue/solution.py
from threading import Condition

class BoundedBlockingQueue(object):

    def __init__(self, capacity: int):
        self.cap = capacity
        self.cond = Condition()
        self.q = deque()
        # Python condition vs lock vs semaphore vs barrier
        '''
        Lock: Mutual exclusion for shared resource. Call lock.lock() while operating on critical section/shared resource. Then call unlock()
        Condition: wait on some condition, blocked until notify() is called
        Semaphore: Limit concurrent access to n threads at a time (Semaphore(n)), block after n
        Barrier: Blocks until all required threads call wait() - i.e. awaits the state where n number of threads have had a chance to 'line up next to the barrier'
        '''

    def enqueue(self, element: int) -> None:
        with self.cond:
            while len(self.q) == self.cap:
                self.cond.wait()
            self.q.append(element)
            self.cond.notifyAll()

    def dequeue(self) -> int:
        with self.cond:
            while len(self.q) == 0:
                self.cond.wait()
            self.cond.notifyAll()
            return self.q.popleft()

    def size(self) -> int:
        return len(self.q)

File: solutions/1209-design-bounded-blocking-queue/solution.java
// import java.util.package;

class BoundedBlockingQueue {

    private final LinkedList<Integer> queue;
    private final int capacity;
    private Lock lock = new ReentrantLock();
    private final Condition queueEmpty = lock.newCondition();
    private final Condition queueFull = lock.newCondition();
    private int size;

    public BoundedBlockingQueue(int capacity) {
        this.capacity = capacity;
        this.queue = new LinkedList<>(); 
        /*
        if (queue == null) {
            synchronized(BoundedBlockingQueue.class) {
                if (queue == null) {
                    queue = new LinkedList<>(); // initialize this only once
                }
            }
        }
        **/
    }
    
    public void enqueue(int element) throws InterruptedException {
        lock.lock();
        try {
            while (size == capacity) {
                queueFull.await();
            }
            queue.offer(element);
            size++;
            queueEmpty.signal();
        } finally {
            lock.unlock();
        }

    }
    
    public int dequeue() throws InterruptedException {
        lock.lock();
        try {
            while (size == 0) {
                queueEmpty.await();
            }
            queueFull.signal();
            size--;
            return queue.poll();
        } finally {
            lock.unlock();
        }
    }
    
    public int size() {
        lock.lock();
        try {
            return size;
        } finally {
            lock.unlock();
        }
    }
}
